File: levelOrder.java
package s30.BFS;

import s30.trees.TreeNode;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;


// Time Complexity O(n); Iterating through all the nodes of tree.
// Space Complexity O(n/2) at max we'll have all leaves in the queue.
public class levelOrder {

    public List<List<Integer>> levelOrder(TreeNode root) {

        List<List<Integer>> res = new ArrayList();

        if(root == null) return res;

        Queue<TreeNode> levelQueue = new LinkedList();

        levelQueue.add(root);


        while(!levelQueue.isEmpty()){


            // Queue<TreeNode> nextLevel = new LinkedList();
            List<Integer> currNodes = new ArrayList();

            int size = levelQueue.size();

            while(size > 0){

                TreeNode node = levelQueue.poll();
                currNodes.add(node.val);

                if(node.left != null) levelQueue.add(node.left);
                if(node.right != null) levelQueue.add(node.right);
                size--;
            }

            res.add(currNodes);

        }


        return res;

    }
    public static void main(String[] args) {

    }
}

File: level_order.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        
#         levels = []
        
#         if not root:
#             return levels
        
#         level = 0
#         queue = deque([root,])
        
#         while queue:
            
#             levels.append([])
            
#             level_length = len(queue)
            
#             for i in range(level_length):
                
#                 node = queue.popleft()
                
#                 levels[level].append(node.val)
                
                
#                 if node.left:
#                     queue.append(node.left)
                    
#                 if node.right:
#                     queue.append(node.right)
                    
#             level += 1
            
#         return levels

        levels = []
    
        if not root:
            return levels
        
        def helper(node, level):
            
            if len(levels) == level:
                levels.append([])
                
            levels[level].append(node.val)
            
            if node.left:
                helper(node.left, level+1)
                
            if node.right:
                helper(node.right, level + 1)
                
        helper(root, 0)
        return levels
                    
        
